flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureReason.java

```diff
@@ -23,60 +23,69 @@
+	CHECKPOINT_SUBSUMED(false, "Checkpoint has been subsumed."),
 
-	CHECKPOINT_DECLINED("Checkpoint was declined."),
+	CHECKPOINT_DECLINED(false, "Checkpoint was declined."),
 
-	CHECKPOINT_DECLINED_TASK_NOT_READY("Checkpoint was declined (tasks not ready)"),
+	CHECKPOINT_DECLINED_TASK_NOT_READY(false, "Checkpoint was declined (tasks not ready)"),
 
-	CHECKPOINT_DECLINED_TASK_NOT_CHECKPOINTING("Task does not support checkpointing"),
+	CHECKPOINT_DECLINED_TASK_NOT_CHECKPOINTING(false, "Task does not support checkpointing"),
 
-	CHECKPOINT_DECLINED_SUBSUMED("Checkpoint was canceled because a barrier from newer checkpoint was received."),
+	CHECKPOINT_DECLINED_SUBSUMED(false, "Checkpoint was canceled because a barrier from newer checkpoint was received."),
 
-	CHECKPOINT_DECLINED_ON_CANCELLATION_BARRIER("Task received cancellation from one of its inputs"),
+	CHECKPOINT_DECLINED_ON_CANCELLATION_BARRIER(false, "Task received cancellation from one of its inputs"),
 
-	CHECKPOINT_DECLINED_ALIGNMENT_LIMIT_EXCEEDED("The checkpoint alignment phase needed to buffer more than the configured maximum bytes"),
+	CHECKPOINT_DECLINED_ALIGNMENT_LIMIT_EXCEEDED(false, "The checkpoint alignment phase needed to buffer more than the configured maximum bytes"),
 
-	CHECKPOINT_DECLINED_INPUT_END_OF_STREAM("Checkpoint was declined because one input stream is finished"),
+	CHECKPOINT_DECLINED_INPUT_END_OF_STREAM(false, "Checkpoint was declined because one input stream is finished"),
 
-	CHECKPOINT_COORDINATOR_SHUTDOWN("CheckpointCoordinator shutdown."),
+	CHECKPOINT_COORDINATOR_SHUTDOWN(false, "CheckpointCoordinator shutdown."),
 
-	CHECKPOINT_COORDINATOR_SUSPEND("Checkpoint Coordinator is suspending."),
+	CHECKPOINT_COORDINATOR_SUSPEND(false, "Checkpoint Coordinator is suspending."),
 
-	JOB_FAILURE("The job has failed."),
+	JOB_FAILURE(false, "The job has failed."),
 
-	JOB_FAILOVER_REGION("FailoverRegion is restarting."),
+	JOB_FAILOVER_REGION(false, "FailoverRegion is restarting."),
 
-	TASK_CHECKPOINT_FAILURE("Task local checkpoint failure."),
+	TASK_CHECKPOINT_FAILURE(false, "Task local checkpoint failure."),
 
-	FINALIZE_CHECKPOINT_FAILURE("Failure to finalize checkpoint."),
+	FINALIZE_CHECKPOINT_FAILURE(false, "Failure to finalize checkpoint."),
 
-	TRIGGER_CHECKPOINT_FAILURE("Trigger checkpoint failure.");
+	TRIGGER_CHECKPOINT_FAILURE(false, "Trigger checkpoint failure.");
 
 	// ------------------------------------------------------------------------
 
+	private final boolean isPreFlight;
 	private final String message;
 
-	CheckpointFailureReason(String message) {
+	CheckpointFailureReason(boolean isPreFlight, String message) {
+		this.isPreFlight = isPreFlight;
 		this.message = message;
 	}
 
 	public String message() {
 		return message;
 	}
+
+	/**
+	 * @return true if this value indicates a failure reason happening before a checkpoint is passed to a job's tasks.
+	 */
+	public boolean isPreFlight() {
+		return isPreFlight;
```

Comment:
consider renaming the field to `preFlight`: https://stackoverflow.com/questions/5322648/for-a-boolean-field-what-is-the-naming-convention-for-its-getter-setter